README.md

```diff
@@ -1,114 +1,168 @@
-```bash
-mkdir build && cd build
-cmake ../libcanard/tests    # Adjust path if necessary
-make
-./run_tests
+The CAN frames generated from the message transfer are now stored in the transmission queue.
+We need to pick them out one by one and have them transmitted.
+Normally, the following fragment should be invoked periodically to unload the CAN frames from the
+prioritized transmission queue into the CAN driver (or several, if redundant interfaces are used):
+
+```c
+for (const CanardFrame* txf = NULL; (txf = canardTxPeek(&ins)) != NULL;)  // Look at the top of the TX queue.
+{
+    if (txf->timestamp_usec < getCurrentMicroseconds())  // Check if the frame has timed out.
+    {
+        if (!pleaseTransmit(&txf))             // Send the frame. Redundant interfaces may be used here.
+        {
+            break;                             // If the driver is busy, break and retry later.
+        }
+    }
+    canardTxPop(&ins);                         // Remove the frame from the queue after it's transmitted.
+    ins.memory_free(&ins, (CanardFrame*)txf);  // Deallocate the dynamic memory afterwards.
+}
 ```
 
-### Coverity Scan
+Transfer reception is done by feeding frames into the transfer reassembly state machine.
+But first, we need to subscribe:
+
+```c
+CanardRxSubscription my_subscription;
+(void) canardRxSubscribe(&ins,
+                         CanardTransferKindResponse,  // Indicate that we want service responses.
+                         123,                         // The Service-ID to subscribe to.
+                         1024,                        // The maximum payload size (max DSDL object size).
+                         &my_subscription);
+```
 
-First, [get the Coverity build tool](https://scan.coverity.com/download?tab=cxx).
-Then build the tests with it:
+We can subscribe and unsubscribe at runtime as many times as we want.
+Normally, however, an embedded application would subscribe once and roll with it.
+Okay, this is how we receive transfers:
+
+```c
+CanardTransfer transfer;
+const int8_t result = canardRxAccept(&ins,
+                                     &received_frame,            // The CAN frame received from the bus.
+                                     redundant_interface_index,  // If the transport is not redundant, use 0.
+                                     &transfer);
+if (result < 0)
+{
+    // An error has occurred: either an argument is invalid or we've ran out of memory.
+    // It is possible to statically prove that an out-of-memory will never occur for a given application if
+    // the heap is sized correctly; for background, refer to the Robson's Proof and the documentation for O1Heap.
+    // Reception of an invalid frame is NOT an error.
+    abort();
+}
+else if (result == 1)
+{
+    processReceivedTransfer(redundant_interface_index, &transfer);  // A transfer has been received, process it.
+    ins.memory_free(&ins, (void*)transfer.payload);  // Deallocate the dynamic memory afterwards.
```

Comment:
why did you decide to have the application manage the dynamic memory?

Reply:
To eliminate unnecessary data copying. There are two instances where the application has to get involved in memory management:

- TX frame processing.
- RX transfer processing.

The first case allows the application to decouple the removal of the frame from the prioritized TX queue from the point where the frame instance is invalidated. Being able to manage the lifetime of the frame (and its payload) allows the application to reduce the amount of data being copied throughout the TX pipeline. If the frame instance was invalidated after popping the queue, some applications would have to copy the data into auxiliary storage first to extend its lifetime, or to avoid popping the queue until the frame is acknowledged by the CAN controller. The latter is hard to implement safely because between the calls to peek() and pop() the top frame of the queue might be changed if the peek-pop sequence is broken by a push().

The second case is this way because in the old libcanard we used to have a callback invoked by the library when a frame is received; the library would then clean things up after the application upon return from the callback. I removed the callback to enhance compatibility with high-reliability software design guidelines which typically discourage the use of function pointers (we still have pointers to malloc/free though but that's a different story and I was actually considering to make them optional so that instead of runtime binding it would be done by the linker; we could revisit this later as a build option). Without the callback, there is no way to let the library auto-free the memory afterward. Besides, even if the callback was kept around the way it was, it would terminate the lifetime of the transfer by the time it is returned from, which may not be compatible with the requirements of the application, requiring it to resort to cloning the data.

Reply:
Thanks for explaining.